### ISSUE TITLE
Changing file name server-side

### ISSUE THREAD

**Issue by Steveb-p** [ankitpokhrel/tus-php#167]
**Is your feature request related to a problem? Please describe.**
I'm looking for ability to customize file names for files uploaded. I've seen code that extracts file names from `Request`
https://github.com/ankitpokhrel/tus-php/blob/3878a4933d9e62d4e9a543e88d2b76fd4787d0e1/src/Tus/Server.php#L344
https://github.com/ankitpokhrel/tus-php/blob/3878a4933d9e62d4e9a543e88d2b76fd4787d0e1/src/Request.php#L121

However, I've noticed that this is reliant on client side code, which is not what I'd like.

**Describe the solution you'd like**
I'd like server to allow attaching a custom function to create file name, and fall back to current behavior.

**Describe alternatives you've considered**
I could use middleware to adjust headers in `Request` and `base64_encode` a new value, but it definitely feels wrong to "fix" request object like that.

**Additional context**
There is also an additional security issue. Since filename is client based and is directly concatenated to final file path it could result in malicious requests replacing or adding files in critical system locations, outside of PHP project directory (not to mention injecting .php files or replacing static assets).

EDIT: Theoretically one could use `tus-server.upload.created` event to overwrite file location, but unfortunately it is called **after** cache is saved :cry:

**Comment by ankitpokhrel** [comment]
@Steveb-p thanks for reporting the security issue, fixed in #168 

To override file location, you can listen to `tus-server.upload.created` where you will have access to the upload key. You can use this key to override the cache content.

**Comment by Steveb-p** [comment]
@ankitpokhrel could we move the emission of `tus-server.upload.created` event before cache is saved? That way there would be no need to manipulate cache in any way, and I don't see any mutation of `File` object from cache saving. All that would need to be done to change filename could be done from inside the event handler interacting with `File` object.

https://github.com/ankitpokhrel/tus-php/blob/66f635531677606ac41d222b764e32f343ab9b11/src/Tus/Server.php#L370-L383

...would be instead...

```php
$file = $this->buildFile([
    'name' => $fileName,
    'offset' => 0,
    'size' => $this->getRequest()->header('Upload-Length'),
    'file_path' => $filePath,
    'location' => $location,
])->setKey($uploadKey)->setChecksum($checksum);
$this->event()->dispatch(
    UploadCreated::NAME,
    new UploadCreated($file, $this->getRequest(), $this->getResponse())
);
$this->cache->set($uploadKey, $file->details() + ['upload_type' => $uploadType]);
```
